File: cocos2dx-js/sdk/AbstractDumper.js
var IDumper = function () {

}

IDumper.prototype.getRoot = function () {
    // :rettype: support.poco.sdk.AbstractNode
}

IDumper.prototype.dumpHierarchy = function () {
    // :rettype: dict or NoneType
}

var AbstractDumper = function () {
    IDumper.call(this)
}
AbstractDumper.prototype = Object.create(IDumper.prototype)

AbstractDumper.prototype.dumpHierarchy = function () {
    return this.dumpHierarchyImpl(this.getRoot())
}

AbstractDumper.prototype.dumpHierarchyImpl = function (node) {
    if (!node) {
        return null
    }

    var payload = node.enumerateAttrs()
    var result = {}
    var children = []
    var nodeChildren = node.getChildren()
    for (var i in nodeChildren) {
        var child = nodeChildren[i]
        if (payload['visible'] || child.getAttr('visible')) {
            children.push(this.dumpHierarchyImpl(child))
        }
    }
    if (children.length > 0) {
        result['children'] = children
    }

    result['name'] = payload['name'] || node.getAttr('name')
    result['payload'] = payload

    return result
}

module.exports = AbstractDumper;

File: cocos2dx-lua/poco/sdk/AbstractDumper.lua
local IDumper = {}
IDumper.__index = IDumper

function IDumper:getRoot() 
    -- :rettype: support.poco.sdk.AbstractNode
end

function IDumper:dumpHierarchy(onlyVisibleNode) 
    -- :rettype: dict or NoneType
end

local AbstractDumper = {}
AbstractDumper.__index = AbstractDumper
setmetatable(AbstractDumper, IDumper)


function AbstractDumper:dumpHierarchy(onlyVisibleNode)
    if onlyVisibleNode == nil then
        onlyVisibleNode = true
    end
    return self:dumpHierarchyImpl(self:getRoot(), onlyVisibleNode)
end


function AbstractDumper:dumpHierarchyImpl(node, onlyVisibleNode)
    if node == nil then
        return nil
    end
    if onlyVisibleNode == nil then
        onlyVisibleNode = true
    end

    local payload = node:enumerateAttrs()
    local result = {}
    local children = {}
    for _, child in ipairs(node:getChildren()) do
        if not onlyVisibleNode or child:getAttr('visible') then
            table.insert(children, self:dumpHierarchyImpl(child, onlyVisibleNode))
        end
    end
    if #children > 0 then
        result['children'] = children
    end
    
    result['name'] = payload['name'] or node:getAttr('name')
    result['payload'] = payload
    
    return result
end

return AbstractDumper